app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt

```diff
@@ -0,0 +1,104 @@
+package com.rockthevote.grommet.ui
+
+import androidx.lifecycle.*
+import com.rockthevote.grommet.data.api.RockyService
+import com.rockthevote.grommet.data.api.model.RegistrationResponse
+import com.rockthevote.grommet.data.db.model.RockyRequest
+import com.rockthevote.grommet.util.DispatcherProvider
+import com.rockthevote.grommet.util.DispatcherProviderImpl
+import kotlinx.coroutines.*
+import retrofit2.adapter.rxjava.Result
+import rx.Observable
+import timber.log.Timber
+
+class MainActivityViewModel(
+    dispatchers: DispatcherProvider = DispatcherProviderImpl(),
+    private val rockyService: RockyService
+) : ViewModel() {
+
+    private val _state = MutableLiveData<MainActivityState>(MainActivityState.Init)
+    val state: LiveData<MainActivityState> = _state
+
+    private val coroutineExceptionHandler = CoroutineExceptionHandler { _, throwable ->
+        // TODO Should we handle this another way? Is it needed?
+        Timber.e(throwable)
+    }
+
+    private val supervisorJob = SupervisorJob()
+
+    private val rockyRequestScope = CoroutineScope(dispatchers.io + coroutineExceptionHandler)
+
+    init {
+        viewModelScope.launch(dispatchers.io) {
+            updateState(MainActivityState.Loading)
+
+            val requests = loadRequestsFromDb()
+
+            val content = MainActivityState.Content(
+                pendingUploads = requests.size,
+                failedUploads = 0 // TODO need to figure out how we're going to calculate this
+            )
+
+            updateState(content)
+        }
+    }
+
+    fun uploadRegistrations() {
+        updateState(MainActivityState.Loading)
```

Comment:
god this is so much cleaner than the old upload service class, mainly due to storing it in a pre-formatted json blob, but still.